### ISSUE TITLE
minetest.get_craft_recipe occasionally returns wrong recipes

### ISSUE THREAD

**Issue by t4im** [minetest/minetest#2222]
`minetest.get_craft_recipe("default:chest")` returns the recipe for the locked chest instead.

**Comment by t4im** [comment]
Just tested, and the issue remains unaffected by #2225

**Comment by t4im** [comment]
A quick unit test for confirmation:
```lua
minetest.after(0, function ()
	local input = minetest.get_craft_recipe("default:chest")
	-- will return default:steel_ingot as the middle item
	assert(input.items[5] ~= "default:steel_ingot")

	input = minetest.get_craft_recipe("default:brick")
	-- not a shapeless recipe, so the third slot should be empty, while the fifth expects the brick
	assert(input.items[3] ~= "default:clay_brick")
	assert(input.items[5] == "default:clay_brick")
end)
```

**Comment by t4im** [comment]
updated issue and unit test, as it appears that only one of the two mentioned cases is an actual issue, the second can be handled with recipe width

**Comment by est31** [comment]
Probable cause: The [check](https://github.com/gregorycu/minetest/blob/master/src/craftdef.cpp#L957) that checks whether a craft recipe from the registered list output matches the output only checks whether the output of the registered recipe starts with the "desired" output.

**Comment by C1ffisme** [comment]
Have you tried this with and without mods?

**Comment by est31** [comment]
From the code it seems `getCraftRecipes` is broken too...

**Comment by t4im** [comment]
fixed with d902bd3
